presto-main/src/test/java/com/facebook/presto/operator/TestHashAggregationOperator.java

```diff
@@ -370,6 +377,101 @@ public void testHashBuilderResize(boolean hashEnabled, long memoryLimitForMerge,
         toPages(operatorFactory, driverContext, input);
     }
 
+    @Test
+    public void testMemoryReservationYield()
+    {
+        // create pages with sizes doubling so that we can enforce a rehash for every new page added
+        List<Integer> hashChannels = Ints.asList(0);
+        RowPagesBuilder rowPagesBuilder = rowPagesBuilder(true, hashChannels, VARCHAR);
+        // query context has GUARANTEED_MEMORY to make sure no query is blocked if it uses memory < 1MB
+        // each hash table entry is of 13 bytes; so we need 1_000_000 / 13 = 77K entries for the first page in order to exceed 1MB limit
+        int length = 1 << 17;
+        int expected = 0;
+        for (int i = 1; i < 5; i++) {
+            expected += length;
+            rowPagesBuilder.addSequencePage(length, length);
+            length = length * 2;
+        }
+        List<Page> input = rowPagesBuilder.build();
+
+        // mock an adjustable memory pool
+        QueryId queryId = new QueryId("test_query");
+        TestMemoryPool memoryPool = new TestMemoryPool();
+        // reserve some memory given free() below can make reservedBytes to be negative
+        memoryPool.reserve(queryId, 1_000_000);
+        QueryContext queryContext = new QueryContext(
+                queryId,
+                new DataSize(256, MEGABYTE),
+                memoryPool,
+                new TestMemoryPool(),
+                executor,
+                scheduledExecutor,
+                new DataSize(256, MEGABYTE),
+                new SpillSpaceTracker(new DataSize(256, MEGABYTE)));
+
+        DriverContext driverContext = createTaskContext(queryContext, executor, TEST_SESSION)
+                .addPipelineContext(0, true, true)
+                .addDriverContext();
+
+        HashAggregationOperatorFactory operatorFactory = new HashAggregationOperatorFactory(
+                0,
+                new PlanNodeId("test"),
+                ImmutableList.of(VARCHAR),
+                hashChannels,
+                ImmutableList.of(),
+                Step.SINGLE,
+                ImmutableList.of(COUNT.bind(ImmutableList.of(0), Optional.empty())),
+                rowPagesBuilder.getHashChannel(),
+                Optional.empty(),
+                1_000,
+                new DataSize(16, MEGABYTE),
+                joinCompiler);
+
+        // test yield by setting memory pool free bytes
+        Operator operator = operatorFactory.createOperator(driverContext);
+        for (Page page : input) {
+            // unblocked
+            assertTrue(operator.needsInput());
+
+            // starting with negative free bytes so that the reservation will fail
+            memoryPool.setFreeBytes(-1);
+
+            operator.addInput(page);
+            // immediately blocked by memory limit
+            assertFalse(operator.needsInput());
```

Comment:
Maybe also check that `operator` did no memory allocations (except try- ones)?

Reply:
The memory will increase a bit given aggregator is increasing in size; the only thing that won't change is the hash table size, which is pretty hard to assert. What I can do here is to assert the increased memory size is not going to be greater than the increased hash table size (that is supposed to happen but not), which is `13 * nextPowerOfTwo((long) (page.getPositionCount() / 0.75))` (where `13` is the size of each entry in the hash table in bytes)

Reply:
Hm... that'd be awfully detailed a test